libutils/logging.c

```diff
@@ -281,6 +285,123 @@ void LogToSystemLog(const char *msg, LogLevel level)
 }
 #endif  /* !__MINGW32__ */
 
+#ifdef HAVE_SYSTEMD_SD_JOURNAL_H
+/**
+ * Count the number of key-value pairs provided in the variable-length
+ * argument list until (but not including) the "MESSAGE" key.
+ */
+static inline size_t CountExtraPairs(va_list va)
+{
+    va_list args;
+    va_copy(args, va);
+
+    size_t count = 0;
+    for (char *key = va_arg(args, char *); !StringEqual(key, "MESSAGE");
+         key = va_arg(args, char *), count++)
+    {
+        va_arg(args, char *);
+    }
+
+    va_end(args);
+    return count;
+}
+
+/**
+ * Add the extra key-value pairs to the iovec.
+ */
+static inline void iovecAddExtraPairs(va_list args,
+                                      struct iovec *const iov,
+                                      const size_t num_pairs)
+{
+    for (size_t i = 0; i < num_pairs; i++)
+    {
+        const char *const key = va_arg(args, char *);
+        const char *const value = va_arg(args, char *);
+
+        char *const pair = StringFormat("%s=%s", key, value);
+        assert(pair != NULL);
+
+        iov[i].iov_base = pair;
+        iov[i].iov_len = strlen(pair);
+    }
+}
+
+/**
+ * Add the "MESSAGE" key-value pair to the iovec.
+ */
+static inline void iovecAddMessagePair(va_list args,
+                                       struct iovec *const iov)
+{
+    NDEBUG_UNUSED const char *const key = va_arg(args, char *);
+    assert(StringEqual(key, "MESSAGE"));
+
+    const char *const format_str = va_arg(args, char *);
+    char *const value = StringVFormat(format_str, args);
+    assert(value != NULL);
+
+    char *const pair = StringFormat("MESSAGE=%s", value);
+    assert(pair != NULL);
+    free(value);
+
+    iov->iov_base = pair;
+    iov->iov_len = strlen(pair);
+}
+
+/**
+ * Add the "PRIORITY" key-value pair to the iovec.
+ */
+static inline void iovecAddPriorityPair(const LogLevel level,
+                                        struct iovec *const iov)
+{
+    const int priority = LogLevelToSyslogPriority(level);
+    char *const pair = StringFormat("PRIORITY=%i", priority);
+    assert(pair != NULL);
+
+    iov->iov_base = pair;
+    iov->iov_len = strlen(pair);
+}
+#endif // HAVE_SYSTEMD_SD_JOURNAL_H
+
+void LogToSystemLogStructured(const LogLevel level, ...)
+{
+    va_list args;
+    va_start(args, level);
+
+#ifdef HAVE_SYSTEMD_SD_JOURNAL_H
+    const size_t num_pairs = CountExtraPairs(args);
+    struct iovec iov[num_pairs + 2]; // +2 for "MESSAGE" and "PRIORITY".
+
+    iovecAddExtraPairs(args, &iov[2], num_pairs);
```

Comment:
> If ap is passed to a function that uses va_arg(ap,type), then the value of ap is undefined after the return of that function.

Generally I think the safest approach here might be to convert the `args` list into a `Seq` and then work with the sequence.

Reply:
Yes, i totally agree, thanks 🚀